src/olympia/amo/middleware.py

```diff
@@ -192,6 +193,28 @@ def process_request(self, request):
             return super(CommonMiddleware, self).process_request(request)
 
 
+class NonAtomicRequestsForSafeHttpMethodsMiddleware(MiddlewareMixin):
+    """
+    Middleware to make the view non-atomic if the HTTP method used is safe,
+    in order to avoid opening and closing a useless transaction.
+    """
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        # This uses undocumented django APIS:
+        # - transaction.get_connection() followed by in_atomic_block property,
+        #   which we need to make sure we're not messing with a transaction
+        #   that has already started (which happens in tests using the regular
+        #   TestCase class)
+        # - _non_atomic_requests(), which set the property to prevent the
+        #   transaction on the view itself. We can't use non_atomic_requests
+        #   (without the '_') as it returns a *new* view, and we can't do that
+        #   in a middleware, we need to modify it in place and return None so
+        #   that the rest of the middlewares are run.
+        is_method_safe = request.method in ('HEAD', 'GET', 'OPTIONS', 'TRACE')
+        if is_method_safe and not transaction.get_connection().in_atomic_block:
+            transaction._non_atomic_requests(view_func, using='default')
+        return None
```

Comment:
Late question: Are the view_func shared by threads? If so, will this work in threading concurrency?

This way, I suspect that POST can occur on a `view_func` in one thread, but `view_func` was marked as non-atomic on another thread handling a GET or HEAD.   

Because comment said that `_non_atomic_requests()` marks the view in-place. So what happens if, lets say, thread1 get a HEAD and thread2 gets a POST? Is it possible for OS to stop thread1(HEAD) between lines 213-214, switch to thread2(POST) up to line 215, switch back to thread1(HEAD) then call `_non_atomic_requests` and switch back to thread2(POST)?

Asking because I want to reuse your middleware as I have the same "issue" of too many transactions to be fixed.

Reply:
Thanks for the comment. It doesn't look thread-safe at all indeed! I've filed https://github.com/mozilla/addons-server/issues/13375 to investigate a fix.